### PR TITLE
Add backupentry and backupbucket controllers for S3-compatible object storage

### LINKED ISSUE #28: Add backupentry and backupbucket controllers for S3-compatible object storage
Add S3 compatible backupbucket and backupentry controllers.

### PR DESCRIPTION
**Pull request by rootfs**
/area backup
/priority normal
/platform kubevirt

**What this PR does / why we need it**:
Add s3 compatible backup controllers
**Which issue(s) this PR fixes**:
Fixes #28 

**Special notes for your reviewer**:

**Release note**:
```improvement operator

```

### CODE REVIEW

File: pkg/s3/s3client.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		},
+	}
+
+	if _, err := c.S3.CreateBucketWithContext(ctx, createBucketInput); err != nil {
+		if aerr, ok := err.(awserr.Error); !ok {
+			return err
+		} else if aerr.Code() != s3.ErrCodeBucketAlreadyExists && aerr.Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+			return err
+		}
+	}
```
Comment (stoyanr) [inline pkg/s3/s3client.go:81]: In the AWS extension, at this point there is code that enables default server side encryption using AES256 algorithm, see https://github.com/gardener/gardener-extension-provider-aws/blob/cd6d41e5ef0528129c1b26a36d8bcdcb66961470/pkg/aws/client/client.go#L386-L400. Why has it been removed here, is it something that is only likely to be supported on AWS itself? Could we put it back and make it configurable perhaps, or check if it's supported and fall back to no encryption if not?

File: pkg/s3/s3client.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			Credentials:      creds,
+			Endpoint:         &endpoint,
+			DisableSSL:       &noSSL,
+			S3ForcePathStyle: &pathStyle,
+		},
+		// Profile: "profile_name",
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create S3 session instance: %v", err)
```
Comment (stoyanr) [inline pkg/s3/s3client.go:57]: The style of wrapping errors used elsewhere in this extension is by using the `github.com/pkg/errors` package.
```suggestion
		return nil, errors.Wrap(err, "could not create S3 session")
```

File: pkg/s3/s3client.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		Bucket: aws.String(bucket),
+		ACL:    aws.String(s3.BucketCannedACLPrivate),
+		CreateBucketConfiguration: &s3.CreateBucketConfiguration{
+			LocationConstraint: aws.String(region),
+		},
+	}
+
+	if _, err := c.S3.CreateBucketWithContext(ctx, createBucketInput); err != nil {
+		if aerr, ok := err.(awserr.Error); !ok {
+			return err
```
Comment (stoyanr) [inline pkg/s3/s3client.go:77]: Let's wrap all errors coming from external packages.
```suggestion
			return errors.Wrap(err, "could not create bucket")
```

File: pkg/s3/s3client.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		} else if aerr.Code() != s3.ErrCodeBucketAlreadyExists && aerr.Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+			return err
```
Comment (stoyanr) [inline pkg/s3/s3client.go:79]: Let's wrap all errors coming from external packages.
```suggestion
			return errors.Wrap(err, "could not create bucket")
```

File: pkg/s3/s3client.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+						DaysAfterInitiation: aws.Int64(7),
+					},
+					Status: aws.String(s3.ExpirationStatusEnabled),
+				},
+			},
+		},
+	}
+
+	_, err := c.S3.PutBucketLifecycleConfigurationWithContext(ctx, putBucketLifecycleConfigurationInput)
+	return err
```
Comment (stoyanr) [inline pkg/s3/s3client.go:104]: Let's wrap all errors coming from external packages.
```suggestion
	return errors.Wrap(err, "could not put bucket lifecycle configuration")
```

File: pkg/s3/s3client.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+// DeleteBucketIfExists deletes the s3 bucket with name <bucket>. If it does not exist,
+// no error is returned.
+func (c *S3Client) DeleteBucketIfExists(ctx context.Context, bucket string) error {
+	if _, err := c.S3.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return nil
+			}
+			if aerr.Code() == "BucketNotEmpty" {
```
Comment (stoyanr) [inline pkg/s3/s3client.go:115]: Could we extract this code as a constant, or use a constant from AWS SDK if available?

File: pkg/s3/s3client.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+				if err := c.DeleteObjectsWithPrefix(ctx, bucket, ""); err != nil {
+					return err
+				}
+				return c.DeleteBucketIfExists(ctx, bucket)
+			}
+		}
+		return err
```
Comment (stoyanr) [inline pkg/s3/s3client.go:122]: Let's wrap all errors coming from external packages.
```suggestion
		return errors.Wrap(err, "could not delete bucket")
```

File: pkg/s3/s3client.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+					Objects: objectIDs,
+					Quiet:   aws.Bool(true),
+				},
+			}); delErr != nil {
+				return false
+			}
+		}
+		return !lastPage
+	}); err != nil {
+		return err
```
Comment (stoyanr) [inline pkg/s3/s3client.go:158]: Let's wrap all errors coming from external packages.
```suggestion
		return errors.Wrap(err, "could not list objects")
```

File: pkg/s3/s3client.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	if delErr != nil {
+		if aerr, ok := delErr.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchKey {
+			return nil
+		}
+		return delErr
```
Comment (stoyanr) [inline pkg/s3/s3client.go:164]: Let's wrap all errors coming from external packages.
```suggestion
		return errors.Wrap(delErr, "could not delete objects")
```

**Review by stoyanr — CHANGES_REQUESTED** [review]
@rootfs Thanks for the PR, it looks quite nice! I have a few suggestions and questions, see the comments, mostly small things. I also have the following concerns:

* As far as I understand, we are basically providing the capability to use any S3-compatible storage (including S3 itself but also e.g. IBM Cloud Object Storage) for backups. None of this is specific to Kubevirt, so is it really correct to implement it in a Kubevirt extension? On the other hand, I don't have any better suggestion, there is no such things as "object store extension" in Gardener and the AWS extension only works with S3 right now. But I would like to verify this with colleagues from the team before approving.
* Have you tested it, and with what storage? How could we obtain credentials to test it with something else than S3 (and even with S3, what would be the "endpoint")?

**Comment by stoyanr** [comment]
@rootfs Taking this PR over as agreed via Slack.
/assign @stoyanr

**Review by stoyanr — APPROVED** [review]

File: pkg/controller/backupbucket/actuator.go [hunk 10]
```diff
@@ -43,12 +45,17 @@ func (a *actuator) InjectClient(client client.Client) error {
 }
 
 func (a *actuator) Reconcile(ctx context.Context, bb *extensionsv1alpha1.BackupBucket) error {
+	bbConfig, err := helper.GetBackupBucketConfig(bb)
```
Comment (rootfs) [inline pkg/controller/backupbucket/actuator.go:48]: why create a separate API for bucket config? It'll require both config, `bucketconfig` as well as s3 secret exist. what about making `EnableEncryption` part of the s3 secret?

Comment (stoyanr) [reply pkg/controller/backupbucket/actuator.go]: This definitely makes sense, thanks. Added it to the secret and removed the API.

**Comment by stoyanr** [comment]
@rootfs I updated the PR as suggested and also added an `insecureSkipVerify` option to the secret so that we could test against ceph storage with self-signed certificate. I tested with AWS as S3 backend. 

I couldn't test with the current ceph storage on the OpenShift cluster as it currently has some issues:
```
aws --endpoint=https://s3-openshift-storage.apps.scp-on-ocp.redhat.dev --no-verify-ssl s3 ls
...

An error occurred (InternalError) when calling the ListBuckets operation (reached max retries: 4): We encountered an internal error. Please try again.
```

If you prefer, we could still merge this PR and test it on OpenShift later.

**Comment by rootfs** [comment]
/lgtm
thanks

**Comment by stoyanr** [comment]
@rootfs I was able to test on OpenShift with the correct endpoint and successfully reconciled backup bucket creation / deletion. The bucket was still not usable by [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore), most likely because it currently ignores the `endpoint` and tries to use the provided credentials with AWS S3. I will work on that in a separate PR in that repo.

**Pull request merged**